### DISCUSSION TITLE
Get current rack utilization in units out of total rack space

### DISCUSSION THREAD

**Discussion by domainzero** [netbox-community/netbox#7332] [Q&A] [answered]
My leadership wants some insight into how many rack units we are using out of the total available rack units we have at our colo.
I was trying to get this information by writing an export template, but it doesn't look like total used rack units are exposed as a data point as there's some calculation that is done to determine the percentage of used space.

Is this something I'll need to write a report for? Is there a simple way to access total utilized rack units out of available rack units in a report?

**Comment by ziggekatten** [comment]
We use a custom script together with the Pandas python library to generate and push an (god forbid) excel file to our intranet containing an aggregated pivot table meeting our needs. Standard export to csv was to limited, but with the power of custom script and Python libraries there is no limit. 

An example, we look at devices and their occupation in Rack units. As soon as they reach 47, they get an discount for a full rack, and additional units are charged separately. 

I spent no more than a day figuring out how to use Pandas for our specific needs. This is one way of doing it, and with alot of power!

**Comment by drmsoffall** [comment, marked as answer]
If I understand your intent, I think you can still do this with export templates if you want. You'll definitely have to iterate over racks to get this info. The trick is working with what we _can_ get; the following is based on the export template example in the docs and a glance at the available attributes on the Rack model.

```
Site,Rack,Occupied,Free,Total{% set ttls = {'u': 0, 'free': 0} %}
{% for rack in queryset %}{% set free_u = rack.get_available_units()|length %}{{ rack.site.name }},{{ rack.name }},{{ rack.u_height - free_u }},{{ free_u }},{{ rack.u_height }}{% set _ = ttls.update({'u': ttls['u']+rack.u_height}) %}{% set _ = ttls.update({'free': ttls['free']+free_u}) %}
{% endfor %}TOTAL,TOTAL,{{ ttls['u'] - ttls['free'] }},{{ ttls['free'] }},{{ ttls['u'] }}
```

Since we can get the total units in a rack (`u_height`) and a list of free units (`get_available_units()`), we can calculate the count of free units and occupied units.  Laid out for CSV-style output but you should be able to edit to another format or exclude information you don't need.

**Reply by domainzero** [reply]
This does exactly what I need! I was trying to call the `get_available_units()` function while messing around earlier, and this definitely puts it in clearer perspective. Thanks!
